### ISSUE TITLE
ZC 1.5.6: Warning notice when add/remove category links in Categories Link Manager

### ISSUE THREAD

**Issue by Pan2020** [zencart/zencart#2311]
When I add or remove categories on a product in Categories Link Manager, the warning appears despite there is master category on a product.

`WARNING: No Master Category is set!`

![NoMasterCategory](https://user-images.githubusercontent.com/21959676/56973963-e920e380-6b3b-11e9-8394-04c59afa0006.png)

This always showed up in yellow bar at the top of admin page every time after I clicked "Update Category Links" regardless whether I updated category links or not. The warning will still trigger even I went straight to click on "Update Category Links" without doing anything!

There is no PHP notice or logged generated for this warning. So I suspect this is more due to certain SQL queries on category link managers that keep triggering this...

**Comment by mc12345678** [comment]
Can confirm that this occurs for linked product.  The code section that seems involved is in `admin/products_to_categories.php` lines 401 through 459.
````php
      // set the linked products master_categories_id product(s)
      for ($i = 0, $n = sizeof($_POST['categories_add']); $i < $n; $i++) {
        // is current master_categories_id in the list?
        if ($zv_check_master_categories_id == true && $_POST['categories_add'][$i] == $current_master_categories_id) {
          $zv_check_master_categories_id = true;
          // array is set above to master category
        } else {
          $zv_check_master_categories_id = false;
          $new_categories_sort_array[] = (int)$_POST['categories_add'][$i];
        }
      }
      // remove existing products_to_categories for current product
      $db->Execute("delete from " . TABLE_PRODUCTS_TO_CATEGORIES . " where products_id='" . $products_filter . "'");
      $reset_master_categories_id = '';
      $old_master_categories_id = $current_master_categories_id;
      // add products to categories in order of master_categories_id first then others
      $verify_current_category_id = false;
      for ($i = 0, $n = sizeof($new_categories_sort_array); $i < $n; $i++) {
        // is current master_categories_id in the list?
        if ($new_categories_sort_array[$i] <= 0) {
          die('I WOULD NOT ADD ' . $new_categories_sort_array[$i] . '<br>');
        } else {
          if ($current_category_id == $new_categories_sort_array[$i]) {
            $verify_current_category_id = true;
          }
          $db->Execute("INSERT INTO " . TABLE_PRODUCTS_TO_CATEGORIES . " (products_id, categories_id)
                        VALUES (" . $products_filter . ", " . (int)$new_categories_sort_array[$i] . ")");
          if ($reset_master_categories_id == '') {
            $reset_master_categories_id = $new_categories_sort_array[$i];
          }
          if ($old_master_categories_id == $new_categories_sort_array[$i]) {
            $reset_master_categories_id = $new_categories_sort_array[$i];
          }
        }
      }
      // reset master_categories_id in products table
      if ($zv_check_master_categories_id == true) {
        // make sure master_categories_id is set to current master_categories_id
        $db->Execute("UPDATE " . TABLE_PRODUCTS . "
                      SET master_categories_id = " . (int)$current_master_categories_id . "
                      WHERE products_id = " . $products_filter);
      } else {
        // reset master_categories_id to current_category_id because it was unselected
        $db->Execute("UPDATE " . TABLE_PRODUCTS . "
                      SET master_categories_id = " . (int)$reset_master_categories_id . "
                      WHERE products_id = " . $products_filter);
      }
      // recalculate price based on new master_categories_id
      zen_update_products_price_sorter($products_filter);
      if ($zv_check_master_categories_id == true) {
        $messageStack->add_session(SUCCESS_MASTER_CATEGORIES_ID, 'success');
      } else {
        $messageStack->add_session(WARNING_MASTER_CATEGORIES_ID, 'warning');
      }
````

There is a logic error in the overall process, though could use some clarification about what really is trying to be done along the way to ensure that the appropriate code/solution is used.

For product that are linked, this section will always set `$zv_check_master_categories_id` to false but will also continue to populate `$new_categories_sort_array` on top of `$new_categories_sort_array[0]` being set to the then value of `$_POST['current_master_categories_id']` which may or may not be a valid/defined `master_categories_id`:
````php
      for ($i = 0, $n = sizeof($_POST['categories_add']); $i < $n; $i++) {
        // is current master_categories_id in the list?
        if ($zv_check_master_categories_id == true && $_POST['categories_add'][$i] == $current_master_categories_id) {
          $zv_check_master_categories_id = true;
          // array is set above to master category
        } else {
          $zv_check_master_categories_id = false;
          $new_categories_sort_array[] = (int)$_POST['categories_add'][$i];
        }
      }
````

This section because we are trying to address "linked" product where there are categories off of the root of the category tree appears to try to prevent adding a product to the top of the tree as seen in the first if statement within the loop. Therefore if the original master_categories_id was 0, then this code would stop further execution at the first array element and prevent any further modification of the linkage regardless if the selections no longer included that category:
````php
        for ($i=0, $n=sizeof($new_categories_sort_array); $i<$n; $i++) {
          // is current master_categories_id in the list?
          if ($new_categories_sort_array[$i] <= 0) {
            die('I WOULD NOT ADD ' . $new_categories_sort_array[$i] . '<br>');
          } else {
            if ($current_category_id == $new_categories_sort_array[$i]) {
              $verify_current_category_id = true;
            }
            $db->Execute("insert into " . TABLE_PRODUCTS_TO_CATEGORIES . "
                    (products_id, categories_id)
                    values (" . $products_filter . ", " . (int)$new_categories_sort_array[$i] . ")");
            if ($reset_master_categories_id == '') {
              $reset_master_categories_id = $new_categories_sort_array[$i];
            }
            if ($old_master_categories_id == $new_categories_sort_array[$i]) {
              $reset_master_categories_id = $new_categories_sort_array[$i];
            }
          }
        }
````
At completion of the above code, if the original products_master_id had still been selected then `$verify_current_category_id` would be set to true and $reset_master_categories_id would be set to the original products_master_id, if not then it would still be false as set before this loop and $reset_master_categories_id would also be equal to the original products_master_id based on being the first array element and no other element meeting the criteria of `$old_master_categories_id == $new_categories_sort_array[$i]`.

This section:
````php
      // reset master_categories_id in products table
      if ($zv_check_master_categories_id == true) {
        // make sure master_categories_id is set to current master_categories_id
        $db->Execute("UPDATE " . TABLE_PRODUCTS . "
                      SET master_categories_id = " . (int)$current_master_categories_id . "
                      WHERE products_id = " . $products_filter);
      } else {
        // reset master_categories_id to current_category_id because it was unselected
        $db->Execute("UPDATE " . TABLE_PRODUCTS . "
                      SET master_categories_id = " . (int)$reset_master_categories_id . "
                      WHERE products_id = " . $products_filter);
      }
````
for a linked product would always update the master_categories_id to be the `$reset_master_categories_id` which as provided above would be the original master_categories_id and not some other category that was in the selection list.

Then in a linked product situation:
````php
      if ($zv_check_master_categories_id == true) {
        $messageStack->add_session(SUCCESS_MASTER_CATEGORIES_ID, 'success');
      } else {
        $messageStack->add_session(WARNING_MASTER_CATEGORIES_ID, 'warning');
      }
````
This would always display that the master categories id was not selected because in the first loop `$zv_check_master_categories_id` is set to false even though/if the master_categories_id for the product had been in the selection list.

And lastly and possibly the only correctly performing portion is whether to return the admin to the same category or to filter the product list to reflect the product in its current category:
````php
      // if product was removed from current categories_id stay in same category
      if (!$verify_current_category_id) {
        zen_redirect(zen_href_link(FILENAME_PRODUCTS_TO_CATEGORIES, 'current_category_id=' . $current_category_id));
      } else {
        zen_redirect(zen_href_link(FILENAME_PRODUCTS_TO_CATEGORIES, 'products_filter=' . $products_filter . '&current_category_id=' . $current_category_id));
      }
````
It seems that PR: https://github.com/zencart/zencart/commit/ae536f0ba55fc79e1ee5601c4906fad860eb857c may have introduced a portion of the problem of above.

I've submitted PR #2325 to fix this issue even though it is beyond the proposed cut-off date of 4/30/19 as indicated in #2296.

**Comment by drbyte** [comment]
Closed by #2325
